distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso

```diff
@@ -0,0 +1,36 @@
+from Standard.Base import Any, Problem_Behavior, Nothing, Error, Panic, Meta, File, File_Format, Plain_Text, Bytes
+from Standard.Base.Error.Common import Unsupported_File_Type_Data, No_Such_Method_Error_Data, Illegal_Argument_Error_Data
+
+from Standard.Table.Delimited.Delimited_Format import Delimited
+from Standard.Table.Excel.Excel_Format import Excel
+
+types = [Plain_Text, Delimited ',', Excel, Bytes]
+
+type Auto_Detect
+    ## If the File_Format supports reading from the file, return a configured instance.
+    can_read : File -> Any
+    can_read self file =
+        reader idx =
+            formatter = types.at idx . can_read file
+            if formatter.is_nothing.not then formatter else
+                if idx + 1 == types.length then Nothing else
+                    @Tail_Call reader (idx + 1)
+        reader 0
+
+    ## PRIVATE
+       Implements the `File.read` for this `File_Format`
+    read : File -> Problem_Behavior -> Any
+    read self file on_problems =
+        reader = self.can_read file
+        if reader == Nothing then Error.throw (Unsupported_File_Type_Data ("No File_Format supports '" + file.extension + "'")) else
+            reader.read file on_problems
+
+    ## PRIVATE
+       Implements the `Table.write` for this `File_Format`.
+    write_table : File -> Table -> Existing_File_Behavior -> Match_Columns -> Problem_Behavior -> Nothing
+    write_table self file table on_existing_file match_columns on_problems =
+        format = self.can_read file
```

Comment:
If in the future `can_read` would want to not only rely on extension but also contents, this logic will not work anymore as in `write` case the file may not yet exist. But for now I guess it's ok, we can always revisit when need arises.